In [23]:
import pandas as pd

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error
from sklearn.feature_selection import RFE
from sklearn.feature_selection import SelectKBest, f_regression

In [24]:
data = pd.read_excel('/home/aleksandr/Документы/GitHub/data/ML6/data_ford_price.xlsx')

In [25]:
data = data[['price','year', 'cylinders', 'odometer', 'lat', 'long', 'weather']]
data.dropna(inplace = True)

In [26]:
y = data['price']
x = data.drop(columns='price')

X_train, X_test, y_train, y_test = train_test_split(x, y, test_size=0.3, random_state=40)

In [27]:
# Выделяем три столбца-признака для обучения, выбранные RFE.
estimator = LinearRegression()
selector_RFE = RFE(estimator, n_features_to_select=3, step=1)
selector_RFE = selector_RFE.fit(X_train, y_train)
 
selector_RFE.get_feature_names_out()

array(['year', 'cylinders', 'lat'], dtype=object)

In [28]:
# Выделяем три столбца-признака для обучения, выбранные SelectKBest.
selector_kBest = SelectKBest(f_regression, k=3)
selector_kBest.fit(X_train, y_train)
 
selector_kBest.get_feature_names_out()

array(['year', 'cylinders', 'odometer'], dtype=object)

In [29]:
# Обучаем регрессия на первых трёх столбцах, оценено качество модели на тесте.
y = data['price']
x = data[['year', 'cylinders', 'odometer']]

X_train, X_test, y_train, y_test = train_test_split(x, y, test_size=0.3, random_state=40)

model = LinearRegression()
model.fit(X_train, y_train)
y_predicted = model.predict(X_test)
 
mae = mean_absolute_error(y_test, y_predicted)
print('MAE: %.3f' % mae)

MAE: 4708.946


In [30]:
# Обучаем регрессия на вторых трёх столбцах, оценено качество модели на тесте.
y = data['price']
x = data[['lat', 'long', 'weather']]

X_train, X_test, y_train, y_test = train_test_split(x, y, test_size=0.3, random_state=40)

model = LinearRegression()
model.fit(X_train, y_train)
y_predicted = model.predict(X_test)
 
mae = mean_absolute_error(y_test, y_predicted)
print('MAE: %.3f' % mae)

MAE: 8812.414


### <center> Вывод

Обучение модели линейной регресси на первых трех признаках, отборанных на основе метода фильтрации (SelectKBest) дает резелттат метрики MAE: 4708,946. В то время как соответствующее обучение на вторых трех признаках, показало резульат метрики MAE: 8812.414. Такая большая разница подтверждает вывод о эффективности применения метода отбора признаков.